File: dat-ruby/test/test_example.rb
# frozen_string_literal: true

require_relative './test_helper'

class TestExamplePorted < Minitest::Test
  def test_issue_and_parse
    dat_manager = Saro::Dat::DatManager.new

    # create certificate
    now = Time.now.to_i
    cert = Saro::Dat::DatCertificate.new(
      0,
      now - 10,
      20,
      1800,
      Saro::Dat::DatSignature.generate(Saro::Dat::DatSignatureAlgorithm::ECDSA_P256),
      Saro::Dat::DatCrypto.generate(Saro::Dat::DatCryptoAlgorithm::IV_AES256_GCM)
    )

    puts cert.exports(false)

    # import certificate
    dat_manager.import_certificates([cert])

    example_plain = "plain text = 평문"
    example_secure = "secure = 암호문"

    begin
      dat = dat_manager.issue(example_plain, example_secure)
      payload = dat_manager.parse(dat)

      assert_equal example_plain, payload.plain
      assert_equal example_secure, payload.secure
      puts "PARSE DAT: #{dat}"
      puts "plain: #{payload.plain}"
      puts "secure: #{payload.secure}"
    rescue OpenSSL::Cipher::CipherError
      # Skip if GCM is not supported
    end
  end

end

File: dat-pypi/tests/test_example.py
import time
import unittest

from saro_dat import DatManager, DatCertificate, DatSignature, DatSignatureAlgorithm, DatCrypto, DatCryptoAlgorithm


class TestExample(unittest.TestCase):


    def test_issue_and_parse(self):
        dat_manager = DatManager()

        # create certificate
        now = int(time.time())
        cert = DatCertificate(0, DatSignature.generate(DatSignatureAlgorithm.ECDSA_P256), DatCrypto.generate(DatCryptoAlgorithm.IV_AES128_GCM), now - 10, now + 10, 1800)

        # import certificate
        dat_manager.import_certificates([cert])

        plain = "Unicode 유니코드 ユニコード 万国码 يونيكود यूनिकोड Юникод 🦄💻"
        secure = "Ciphertext 암호문 暗号文 密文 Шифротекст Texte chiffré Geheimtext نص مشفر सिफरपाठ 🔐"

        dat = dat_manager.issue(plain, secure)
        payload = dat_manager.parse(dat)

        assert payload.plain == plain
        assert payload.secure == secure
        print(f"PARSE DAT: {dat}")
        print(f"plain: {payload.plain}")
        print(f"secure: {payload.secure}")

        


if __name__ == "__main__":
    unittest.main()


